src/java/htsjdk/samtools/SAMBinaryTagAndValue.java

```diff
@@ -73,6 +74,34 @@ private boolean typeSafeEquals(final SAMBinaryTagAndValue that) {
         }
     }
 
+    private boolean valueEquals(SAMBinaryTagAndValue that) {
+        if (this.value == null) {
+            return that.value == null;
+        }
+        else if (that.value == null) {
+            return false;
+        }
+        else if (this.value instanceof byte[]) {
+            return that.value instanceof byte[] ?
+                Arrays.equals((byte[])this.value, (byte[])that.value) : false;
+        }
+        else if (this.value instanceof short[]) {
+            return that.value instanceof short[] ?
+                    Arrays.equals((short[])this.value, (short[])that.value) : false;
+        }
+        else if (this.value instanceof int[]) {
+            return that.value instanceof int[] ?
+                    Arrays.equals((int[])this.value, (int[])that.value) : false;
+        }
+        else if (this.value instanceof float[]) {
+            return that.value instanceof float[] ?
+                    Arrays.equals((float[])this.value, (float[])that.value) : false;
+        }
+        else {
+            return this.value.equals(that.value);
```

Comment:
Those cases are all prohibited (for *SAMRecords* anyway) by the [isAllowedAttributeValue](https://github.com/samtools/htsjdk/blob/master/src/java/htsjdk/samtools/SAMRecord.java#L1397) guard. That should really be in SAMBinaryTagAndValue since its a public class. I wonder if we should move it in there (which could break backward compatibility for some uses), since I don't see a way to keep the copy/equals semantics that we want for SAMRecord attributes if SAMBinaryTagAndValue can take any object as an attribute.